monai/transforms/transform.py

```diff
@@ -118,6 +120,79 @@ def _log_stats(data, prefix: Optional[str] = "Data"):
         raise RuntimeError(f"applying transform {transform}") from e
 
 
+class ILazyTransform:
+    """
+    An interface to indicate that the transform has the capability to describe
+    its operation as an affine matrix or grid with accompanying metadata. This
+    interface can be extended from by people adapting transforms to the MONAI framework as well as
+    by implementors of MONAI transforms.
+    """
+
+    @property
+    def lazy_evaluation(
+            self,
+    ):
+        """
+        Get whether lazy_evaluation is enabled for this transform instance.
+
+        Returns:
+            True if the transform is operating in a lazy fashion, False if not.
+        """
+        raise NotImplementedError()
+
+    @lazy_evaluation.setter
+    def lazy_evaluation(
+            self,
+            enabled: bool
+    ):
+        """
+        Set whether lazy_evaluation is enabled for this transform instance.
+
+        Args:
+            enabled: True if the transform should operate in a lazy fashion, False if not.
+        """
+        raise NotImplementedError()
+
+
+class IRandomizableTransform:
```

Comment:
I understand that we need `IMultiSampleTransform`, but I'm not sure about `IRandomizableTransform` or `ILazyTransform`, these are not required concepts at the moment according to the prototypes... I can redirect my PR (https://github.com/Project-MONAI/MONAI/pull/5407) to your fork if that addresses any potential merging conflicts

Reply:
If you look at the dictionary transform implementations on #4922, they make use of IRandomizableTransform.

The point of IRandomizableTransform is twofold:
1. It can be the way that a transform is detected as randomizable or not, regardless of whether it is implemented in MONAI or extended from another library
2. Randomizable dictionary transforms inherit from RandomizableTransform but then create a randomizable array transform to do the heavy lifting. However, they only delegate _most_ of the randomization to the array class, whereas
they could delegate all of it. IRandomizableTransform would allow that to be fixed as the dictionary transform can just say "hey, I'm randomizable" and then delegate all of its work to the array transform rather than having an implementation that is only partially used

The addition of this interface is partially motivated by the way I've implemented RandRotated in the big PR #4922 .

https://github.com/Project-MONAI/MONAI/blob/62f8172f020c3c8b6ac182b6b4ca5c3805f9877b/monai/transforms/atmostonce/dictionary.py#L174

I might have gone too far in creating a separate Randomizer class that can be reused across the dictionary and array implementations, but I think it results in pretty clean code.

Reply:
I see, that looks like another refactoring effort introducing new concepts of `Randomizer` and `IRandomizableTransform`... how about we limit the scope for lazy resampling feature, and all the other nice-to-have features stay on a feature branch for now? 

really hope we can ship the lazy resampling soon. The other refactorings add much risks but at the moment they are not critical from a user experience's perspective...